objstore.go

```diff
@@ -593,12 +603,23 @@ func (b *metricBucket) Upload(ctx context.Context, name string, r io.Reader) err
 	b.ops.WithLabelValues(op).Inc()
 
 	start := time.Now()
-	if err := b.bkt.Upload(ctx, name, r); err != nil {
+	err := b.bkt.Upload(ctx, name, r)
+	if err != nil {
 		if !b.isOpFailureExpected(err) && ctx.Err() != context.Canceled {
 			b.opsFailures.WithLabelValues(op).Inc()
 		}
 		return err
 	}
+
+	rc := &nopCloserWithObjectSize{
+		Reader: r,
+	}
+
+	rc.Close()
+	if writtenBytes, err := rc.ObjectSize(); err != nil {
```

Comment:
How about we fetch the object size only after the upload? This way we will only perform this operation if the upload was indeed successful, sparing us from having to compute the size without needing it.

Reply:
I agree with that part that it should only count successful uploads with the histogram, but I don't think we need to use a method like `.ObjectSize()`. 

Let look at an example, how the method would be used:

```
// program uploads something
reader := bytes.NewReader([]byte("my-file-content")
Upload(ctx, "my-file.txt", reader)
```

Then it would end up in the `metricsBucket.Upload` which in turn calls `provider.Upload`.

Then in the provider implementation it would retrieve the content by calling `reader.Read()` at least once, but potentially multiple times.

This `Read()` method returns each time `n int,err error`, and `n` is the amount of bytes read, from the reader. So to get the uploaded size we need to sum up the `n`s returned by `Read` which will be the exact amount of bytes that have been uploaded.

Reply:
Sorry I was not clear on my part, I used the words object size but I did not mean the Object size function but the size of the object uploaded 😊 

Will apply these changes as discussed.